### PR TITLE
PostgreSQL attribute update and various small improvements

### PR DESCRIPTION
**Pull request by dennishendriksen**

### CODE REVIEW

File: molgenis-data-postgresql/src/main/java/org/molgenis/data/postgresql/PostgreSqlQueryGenerator.java [hunk 1]
```diff
@@ -475,15 +471,15 @@ private static String getSqlDropTable(String tableName)
...
+						result.append(' ').append(columnName);
 					}
 					else
 					{
 						result.append(" CAST(").append(columnName).append(" as TEXT)");
 					}
 
 					result.append(" LIKE ?");
-					parameters.add("%" + PostgreSqlUtils.getPostgreSqlQueryValue(r.getValue(), attr) + "%");
+					parameters.add("%" + PostgreSqlUtils.getPostgreSqlQueryValue(r.getValue(), attr) + '%');
```
Comment (Mark-de-Haan) [inline molgenis-data-postgresql/src/main/java/org/molgenis/data/postgresql/PostgreSqlQueryGenerator.java]: Shouldnt the first % also be with single quotes?

Comment (dennishendriksen) [inline molgenis-data-postgresql/src/main/java/org/molgenis/data/postgresql/PostgreSqlQueryGenerator.java]: Then you would try to add a String to a char, which won't compile

File: molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java [hunk 2]
```diff
@@ -364,37 +364,44 @@ private void validateUpdate(AttributeMetaData currentAttr, AttributeMetaData new
...
 		// not implemented: HTML -> CATEGORICAL, CATEGORICAL_MREF, EMAIL, ENUM, HYPERLINK, MREF, XREF
 		// not allowed    : HTML -> COMPOUND, FILE, HTML
-		DATA_TYPE_ALLOWED_TRANSITIONS
-				.put(HTML, EnumSet.of(BOOL, DATE, DATE_TIME, DECIMAL, INT, LONG, SCRIPT, STRING, TEXT));
+		DATA_TYPE_ALLOWED_TRANSITIONS.put(HTML,
+				EnumSet.of(CATEGORICAL, BOOL, DATE, DATE_TIME, DECIMAL, INT, LONG, SCRIPT, STRING, TEXT, XREF));
 		DATA_TYPE_ALLOWED_TRANSITIONS.put(HYPERLINK, EnumSet.of(SCRIPT, HTML, STRING, TEXT));
-		DATA_TYPE_ALLOWED_TRANSITIONS.put(INT, EnumSet.of(DECIMAL, HTML, LONG, SCRIPT, TEXT));
-		DATA_TYPE_ALLOWED_TRANSITIONS.put(LONG, EnumSet.of(DECIMAL, HTML, INT, SCRIPT, TEXT));
+		DATA_TYPE_ALLOWED_TRANSITIONS.put(INT, EnumSet.of(CATEGORICAL, DECIMAL, HTML, LONG, SCRIPT, TEXT, XREF));
```
Comment (Mark-de-Haan) [inline molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java]: Missing String as allowed transition

File: molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java [hunk 3]
```diff
@@ -364,37 +364,44 @@ private void validateUpdate(AttributeMetaData currentAttr, AttributeMetaData new
...
+		DATA_TYPE_ALLOWED_TRANSITIONS.put(LONG, EnumSet.of(CATEGORICAL, DECIMAL, HTML, INT, SCRIPT, TEXT, XREF));
```
Comment (Mark-de-Haan) [inline molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java]: Missing String as allowed transition

File: molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java [hunk 4]
```diff
@@ -364,37 +364,44 @@ private void validateUpdate(AttributeMetaData currentAttr, AttributeMetaData new
...
+		DATA_TYPE_ALLOWED_TRANSITIONS.put(CATEGORICAL,
+				EnumSet.of(BOOL, DATE, DATE_TIME, DECIMAL, HTML, INT, LONG, SCRIPT, STRING, TEXT, XREF));
 		DATA_TYPE_ALLOWED_TRANSITIONS.put(CATEGORICAL_MREF, EnumSet.of(MREF));
 		DATA_TYPE_ALLOWED_TRANSITIONS.put(COMPOUND, EnumSet.noneOf(AttributeType.class));
-		DATA_TYPE_ALLOWED_TRANSITIONS.put(DATE, EnumSet.of(DATE_TIME, HTML, SCRIPT, STRING, TEXT));
-		DATA_TYPE_ALLOWED_TRANSITIONS.put(DATE_TIME, EnumSet.of(DATE, HTML, SCRIPT, STRING, TEXT));
-		DATA_TYPE_ALLOWED_TRANSITIONS.put(DECIMAL, EnumSet.of(HTML, INT, LONG, SCRIPT, TEXT));
+		DATA_TYPE_ALLOWED_TRANSITIONS.put(DATE, EnumSet.of(CATEGORICAL, DATE_TIME, HTML, SCRIPT, STRING, TEXT, XREF));
+		DATA_TYPE_ALLOWED_TRANSITIONS.put(DATE_TIME, EnumSet.of(CATEGORICAL, DATE, HTML, SCRIPT, STRING, TEXT, XREF));
+		DATA_TYPE_ALLOWED_TRANSITIONS.put(DECIMAL, EnumSet.of(CATEGORICAL, HTML, INT, LONG, SCRIPT, TEXT, XREF));
```
Comment (Mark-de-Haan) [inline molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java]: Also missing string

File: molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java [hunk 5]
```diff
@@ -364,37 +364,44 @@ private void validateUpdate(AttributeMetaData currentAttr, AttributeMetaData new
...
 		// not implemented: TEXT -> CATEGORICAL, CATEGORICAL_MREF, EMAIL, ENUM, HYPERLINK, MREF, XREF
 		// not allowed    : TEXT -> COMPOUND, FILE, TEXT
 		// not implemented: STRING -> CATEGORICAL, CATEGORICAL_MREF, EMAIL, ENUM, HYPERLINK, MREF, XREF
 		// not allowed    : STRING -> COMPOUND, FILE, STRING
-		DATA_TYPE_ALLOWED_TRANSITIONS
-				.put(STRING, EnumSet.of(BOOL, DATE, DATE_TIME, DECIMAL, HTML, INT, LONG, SCRIPT, TEXT));
-		DATA_TYPE_ALLOWED_TRANSITIONS
-				.put(TEXT, EnumSet.of(BOOL, DATE, DATE_TIME, DECIMAL, HTML, INT, LONG, SCRIPT, STRING));
-		DATA_TYPE_ALLOWED_TRANSITIONS.put(XREF, EnumSet.of(CATEGORICAL));
+		DATA_TYPE_ALLOWED_TRANSITIONS.put(STRING,
```
Comment (Mark-de-Haan) [inline molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java]: TypeTest: When I want to change xstring type to an XREF I get a nullpointer:

java.lang.NullPointerException: null
	at org.molgenis.data.postgresql.PostgreSqlQueryGenerator.getPostgreSqlType(PostgreSqlQueryGenerator.java:811) ~[molgenis-data-postgresql-2.0.0-SNAPSHOT.jar:na]
	at org.molgenis.data.postgresql.PostgreSqlQueryGenerator.getSqlSetDataType(PostgreSqlQueryGenerator.java:90) ~[molgenis-data-postgresql-2.0.0-SNAPSHOT.jar:na]

File: molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java [hunk 6]
```diff
@@ -364,37 +364,44 @@ private void validateUpdate(AttributeMetaData currentAttr, AttributeMetaData new
...
-		DATA_TYPE_ALLOWED_TRANSITIONS
-				.put(HTML, EnumSet.of(BOOL, DATE, DATE_TIME, DECIMAL, INT, LONG, SCRIPT, STRING, TEXT));
+		DATA_TYPE_ALLOWED_TRANSITIONS.put(HTML,
+				EnumSet.of(CATEGORICAL, BOOL, DATE, DATE_TIME, DECIMAL, INT, LONG, SCRIPT, STRING, TEXT, XREF));
 		DATA_TYPE_ALLOWED_TRANSITIONS.put(HYPERLINK, EnumSet.of(SCRIPT, HTML, STRING, TEXT));
-		DATA_TYPE_ALLOWED_TRANSITIONS.put(INT, EnumSet.of(DECIMAL, HTML, LONG, SCRIPT, TEXT));
-		DATA_TYPE_ALLOWED_TRANSITIONS.put(LONG, EnumSet.of(DECIMAL, HTML, INT, SCRIPT, TEXT));
+		DATA_TYPE_ALLOWED_TRANSITIONS.put(INT, EnumSet.of(CATEGORICAL, DECIMAL, HTML, LONG, SCRIPT, TEXT, XREF));
+		DATA_TYPE_ALLOWED_TRANSITIONS.put(LONG, EnumSet.of(CATEGORICAL, DECIMAL, HTML, INT, SCRIPT, TEXT, XREF));
 		DATA_TYPE_ALLOWED_TRANSITIONS.put(MREF, EnumSet.of(CATEGORICAL_MREF));
```
Comment (Mark-de-Haan) [inline molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java]: TypeTest: When I Change xmref to xmref_categorical 

```
java.lang.RuntimeException: Illegal attribute type [CATEGORICAL_MREF]
	at org.molgenis.data.postgresql.PostgreSqlQueryGenerator.getPostgreSqlType(PostgreSqlQueryGenerator.java:835) ~[molgenis-data-postgresql-2.0.0-SNAPSHOT.jar:na]
	at org.molgenis.data.postgresql.PostgreSqlQueryGenerator.getSqlSetDataType(PostgreSqlQueryGenerator.java:90) ~[molgenis-data-postgresql-2.0.0-SNAPSHOT.jar:na]
```

File: molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java [hunk 7]
```diff
@@ -364,37 +364,44 @@ private void validateUpdate(AttributeMetaData currentAttr, AttributeMetaData new
 	static
 	{
 		DATA_TYPE_ALLOWED_TRANSITIONS = new EnumMap<>(AttributeType.class);
-		DATA_TYPE_ALLOWED_TRANSITIONS.put(BOOL, EnumSet.of(DECIMAL, INT, LONG, SCRIPT, STRING, TEXT));
-		DATA_TYPE_ALLOWED_TRANSITIONS.put(CATEGORICAL, EnumSet.of(XREF));
+		DATA_TYPE_ALLOWED_TRANSITIONS
+				.put(BOOL, EnumSet.of(CATEGORICAL, DECIMAL, INT, LONG, SCRIPT, STRING, TEXT, XREF));
```
Comment (Mark-de-Haan) [inline molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java]: TypeTest: xbool to type categorical -> nullpointer

```
java.lang.NullPointerException: null
	at org.molgenis.data.postgresql.PostgreSqlQueryGenerator.getPostgreSqlType(PostgreSqlQueryGenerator.java:811) ~[molgenis-data-postgresql-2.0.0-SNAPSHOT.jar:na]
	at org.molgenis.data.postgresql.PostgreSqlQueryGenerator.getSqlSetDataType(PostgreSqlQueryGenerator.java:90) ~[molgenis-data-postgresql-2.0.0-SNAPSHOT.jar:na]
```

File: molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java (hunk 2 above)
Comment (Mark-de-Haan) [inline molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java]: INT back to bool?

File: molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java (hunk 7 above)
Comment (Mark-de-Haan) [inline molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java]: bool -> xref: nullpointer

File: molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java (hunk 5 above)
Comment (dennishendriksen) [inline molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java]: You forget to select a refEntity, I'll update the validation.

Comment (Mark-de-Haan) [inline molgenis-data/src/main/java/org/molgenis/data/meta/AttributeMetaDataRepositoryDecorator.java]: Selecting a RefEntity in the forms is not possible, it is grayed out

**Pull request merged**